Daemon crash-looping (logs attached)

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpx/_client.py", line 1730, in _send_single_request
    response = await transport.handle_async_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpx/_transports/default.py", line 394, in handle_async_request
    resp = await self._pool.handle_async_request(req)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpcore/_async/connection_pool.py", line 256, in handle_async_request
    raise exc from None
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpcore/_async/connection_pool.py", line 236, in handle_async_request
    response = await connection.handle_async_request(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        pool_request.request
        ^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpcore/_async/connection.py", line 101, in handle_async_request
    raise exc
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpcore/_async/connection.py", line 78, in handle_async_request
    stream = await self._connect(request)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpcore/_async/connection.py", line 124, in _connect
    stream = await self._network_backend.connect_tcp(**kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpcore/_backends/auto.py", line 31, in connect_tcp
    return await self._backend.connect_tcp(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<5 lines>...
    )
    ^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpcore/_backends/anyio.py", line 115, in connect_tcp
    stream: anyio.abc.ByteStream = await anyio.connect_tcp(
                                   ^^^^^^^^^^^^^^^^^^^^^^^^
    ...<3 lines>...
    )
    ^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/anyio/_core/_sockets.py", line 215, in connect_tcp
    gai_res = await getaddrinfo(
              ^^^^^^^^^^^^^^^^^^
        target_host, remote_port, family=family, type=socket.SOCK_STREAM
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/anyio/_core/_sockets.py", line 649, in getaddrinfo
    gai_res = await get_async_backend().getaddrinfo(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        encoded_host, port, family=family, type=type, proto=proto, flags=flags
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/anyio/_backends/_asyncio.py", line 2732, in getaddrinfo
    return await get_running_loop().getaddrinfo(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        host, port, family=family, type=type, proto=proto, flags=flags
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/opt/homebrew/Cellar/python@3.14/3.14.3_1/Frameworks/Python.framework/Versions/3.14/lib/python3.14/asyncio/base_events.py", line 936, in getaddrinfo
    return await self.run_in_executor(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
        None, getaddr_func, host, port, family, type, proto, flags)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
asyncio.exceptions.CancelledError: Cancelled via cancel scope 109909490 by <Task pending name='Task-4' coro=<<async_generator_athrow without __name__>()>>
2026-03-04 16:13:13,173 [INFO] corphish.daemon: Daemon started, listening on chat 8757952223
2026-03-04 16:13:23,500 [INFO] httpx: HTTP Request: POST https://api.telegram.org/bot8645862118:AAETB7EJPgjPc2GdSq4gPKvH17QqiUFvt68/getUpdates "HTTP/1.1 200 OK"
2026-03-04 16:13:25,733 [INFO] corphish.daemon: Daemon started, listening on chat 8757952223
2026-03-04 16:13:36,205 [INFO] httpx: HTTP Request: POST https://api.telegram.org/bot8645862118:AAETB7EJPgjPc2GdSq4gPKvH17QqiUFvt68/getUpdates "HTTP/1.1 200 OK"
2026-03-04 16:13:45,870 [INFO] httpx: HTTP Request: POST https://api.telegram.org/bot8645862118:AAETB7EJPgjPc2GdSq4gPKvH17QqiUFvt68/getUpdates "HTTP/1.1 200 OK"
2026-03-04 16:13:45,874 [INFO] corphish.daemon: [user] Hello again
2026-03-04 16:13:45,874 [INFO] claude_agent_sdk._internal.transport.subprocess_cli: Using bundled Claude Code CLI: /Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/claude_agent_sdk/_bundled/claude
2026-03-04 16:13:50,051 [INFO] corphish.daemon: [assistant] Hello. What can I help you with?
2026-03-04 16:13:50,055 [ERROR] asyncio: unhandled exception during asyncio.run() shutdown
task: <Task finished name='Task-4' coro=<<async_generator_athrow without __name__>()> exception=RuntimeError('Attempted to exit cancel scope in a different task than it was entered in')>
Traceback (most recent call last):
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/claude_agent_sdk/_internal/client.py", line 143, in process_query
    yield message
GeneratorExit

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/claude_agent_sdk/_internal/client.py", line 146, in process_query
    await query.close()
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/claude_agent_sdk/_internal/query.py", line 622, in close
    await self._tg.__aexit__(None, None, None)
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/anyio/_backends/_asyncio.py", line 789, in __aexit__
    if self.cancel_scope.__exit__(type(exc), exc, exc.__traceback__):
       ~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/anyio/_backends/_asyncio.py", line 461, in __exit__
    raise RuntimeError(
    ...<2 lines>...
    )
RuntimeError: Attempted to exit cancel scope in a different task than it was entered in
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/Users/zhlhan/.local/share/corphish/corphish/__main__.py", line 33, in <module>
    main()
    ~~~~^^
  File "/Users/zhlhan/.local/share/corphish/corphish/__main__.py", line 29, in main
    asyncio.run(_async_main())
    ~~~~~~~~~~~^^^^^^^^^^^^^^^
  File "/opt/homebrew/Cellar/python@3.14/3.14.3_1/Frameworks/Python.framework/Versions/3.14/lib/python3.14/asyncio/runners.py", line 204, in run
    return runner.run(main)
           ~~~~~~~~~~^^^^^^
  File "/opt/homebrew/Cellar/python@3.14/3.14.3_1/Frameworks/Python.framework/Versions/3.14/lib/python3.14/asyncio/runners.py", line 127, in run
    return self._loop.run_until_complete(task)
           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^^^^^
  File "/opt/homebrew/Cellar/python@3.14/3.14.3_1/Frameworks/Python.framework/Versions/3.14/lib/python3.14/asyncio/base_events.py", line 719, in run_until_complete
    return future.result()
           ~~~~~~~~~~~~~^^
  File "/Users/zhlhan/.local/share/corphish/corphish/__main__.py", line 24, in _async_main
    await dispatch(args)
  File "/Users/zhlhan/.local/share/corphish/corphish/cli.py", line 210, in dispatch
    await run_daemon()
  File "/Users/zhlhan/.local/share/corphish/corphish/daemon.py", line 110, in run_daemon
    await send_message_fn(bot, chat_id, reply)
  File "/Users/zhlhan/.local/share/corphish/corphish/chat.py", line 54, in send_message
    return await bot.send_message(chat_id=chat_id, text=text)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/telegram/_bot.py", line 1122, in send_message
    return await self._send_message(
           ^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<21 lines>...
    )
    ^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/telegram/_bot.py", line 819, in _send_message
    result = await self._post(
             ^^^^^^^^^^^^^^^^^
    ...<7 lines>...
    )
    ^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/telegram/_bot.py", line 703, in _post
    return await self._do_post(
           ^^^^^^^^^^^^^^^^^^^^
    ...<6 lines>...
    )
    ^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/telegram/_bot.py", line 732, in _do_post
    result = await request.post(
             ^^^^^^^^^^^^^^^^^^^
    ...<6 lines>...
    )
    ^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/telegram/request/_baserequest.py", line 198, in post
    result = await self._request_wrapper(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<7 lines>...
    )
    ^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/telegram/request/_baserequest.py", line 305, in _request_wrapper
    code, payload = await self.do_request(
                    ^^^^^^^^^^^^^^^^^^^^^^
    ...<7 lines>...
    )
    ^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/telegram/request/_httpxrequest.py", line 279, in do_request
    res = await self._client.request(
          ^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<6 lines>...
    )
    ^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpx/_client.py", line 1540, in request
    return await self.send(request, auth=auth, follow_redirects=follow_redirects)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpx/_client.py", line 1629, in send
    response = await self._send_handling_auth(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<4 lines>...
    )
    ^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpx/_client.py", line 1657, in _send_handling_auth
    response = await self._send_handling_redirects(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<3 lines>...
    )
    ^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpx/_client.py", line 1694, in _send_handling_redirects
    response = await self._send_single_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpx/_client.py", line 1730, in _send_single_request
    response = await transport.handle_async_request(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpx/_transports/default.py", line 394, in handle_async_request
    resp = await self._pool.handle_async_request(req)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpcore/_async/connection_pool.py", line 256, in handle_async_request
    raise exc from None
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpcore/_async/connection_pool.py", line 236, in handle_async_request
    response = await connection.handle_async_request(
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        pool_request.request
        ^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpcore/_async/connection.py", line 101, in handle_async_request
    raise exc
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpcore/_async/connection.py", line 78, in handle_async_request
    stream = await self._connect(request)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpcore/_async/connection.py", line 124, in _connect
    stream = await self._network_backend.connect_tcp(**kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpcore/_backends/auto.py", line 31, in connect_tcp
    return await self._backend.connect_tcp(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    ...<5 lines>...
    )
    ^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/httpcore/_backends/anyio.py", line 115, in connect_tcp
    stream: anyio.abc.ByteStream = await anyio.connect_tcp(
                                   ^^^^^^^^^^^^^^^^^^^^^^^^
    ...<3 lines>...
    )
    ^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/anyio/_core/_sockets.py", line 215, in connect_tcp
    gai_res = await getaddrinfo(
              ^^^^^^^^^^^^^^^^^^
        target_host, remote_port, family=family, type=socket.SOCK_STREAM
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/anyio/_core/_sockets.py", line 649, in getaddrinfo
    gai_res = await get_async_backend().getaddrinfo(
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        encoded_host, port, family=family, type=type, proto=proto, flags=flags
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/Users/zhlhan/.local/share/corphish/.venv/lib/python3.14/site-packages/anyio/_backends/_asyncio.py", line 2732, in getaddrinfo
    return await get_running_loop().getaddrinfo(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        host, port, family=family, type=type, proto=proto, flags=flags
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/opt/homebrew/Cellar/python@3.14/3.14.3_1/Frameworks/Python.framework/Versions/3.14/lib/python3.14/asyncio/base_events.py", line 936, in getaddrinfo
    return await self.run_in_executor(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
        None, getaddr_func, host, port, family, type, proto, flags)
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
asyncio.exceptions.CancelledError: Cancelled via cancel scope 10b471790 by <Task pending name='Task-4' coro=<<async_generator_athrow without __name__>()>>
2026-03-04 16:13:50,388 [INFO] corphish.daemon: Daemon started, listening on chat 8757952223
2026-03-04 16:14:00,706 [INFO] httpx: HTTP Request: POST https://api.telegram.org/bot8645862118:AAETB7EJPgjPc2GdSq4gPKvH17QqiUFvt68/getUpdates "HTTP/1.1 200 OK"
2026-03-04 16:14:11,840 [INFO] httpx: HTTP Request: POST https://api.telegram.org/bot8645862118:AAETB7EJPgjPc2GdSq4gPKvH17QqiUFvt68/getUpdates "HTTP/1.1 200 OK"
2026-03-04 16:14:23,002 [INFO] httpx: HTTP Request: POST https://api.telegram.org/bot8645862118:AAETB7EJPgjPc2GdSq4gPKvH17QqiUFvt68/getUpdates "HTTP/1.1 200 OK"
2026-03-04 16:14:34,556 [INFO] httpx: HTTP Request: POST https://api.telegram.org/bot8645862118:AAETB7EJPgjPc2GdSq4gPKvH17QqiUFvt68/getUpdates "HTTP/1.1 200 OK"
2026-03-04 16:14:45,735 [INFO] httpx: HTTP Request: POST https://api.telegram.org/bot8645862118:AAETB7EJPgjPc2GdSq4gPKvH17QqiUFvt68/getUpdates "HTTP/1.1 200 OK"
2026-03-04 16:14:56,896 [INFO] httpx: HTTP Request: POST https://api.telegram.org/bot8645862118:AAETB7EJPgjPc2GdSq4gPKvH17QqiUFvt68/getUpdates "HTTP/1.1 200 OK"
2026-03-04 16:15:08,057 [INFO] httpx: HTTP Request: POST https://api.telegram.org/bot8645862118:AAETB7EJPgjPc2GdSq4gPKvH17QqiUFvt68/getUpdates "HTTP/1.1 200 OK"

--- a/corphish/claude_client.py
+++ b/corphish/claude_client.py
@@ -140,4 +140,10 @@ class ClaudeClient:
                     if message.result:
                         return message.result
 
+        # Yield control so any pending anyio cancel-scope cleanup from the
+        # SDK's internal task group settles before the caller does further I/O.
+        # Without this, a leaked cancellation can fire on the next await
+        # (e.g. send_message), causing CancelledError.  See issue #33.
+        await asyncio.sleep(0)
+
         return last_text

--- a/corphish/daemon.py
+++ b/corphish/daemon.py
@@ -103,6 +103,12 @@ async def run_daemon(
             except Exception:
                 logger.exception("Claude call failed for message: %s", user_text)
                 continue
+            except asyncio.CancelledError:
+                logger.warning(
+                    "Claude call cancelled (SDK cleanup leak) for message: %s",
+                    user_text,
+                )
+                continue
 
             logger.info("[assistant] %s", reply)
 
@@ -111,6 +117,11 @@ async def run_daemon(
             except Exception:
                 logger.exception(
                     "Failed to send reply via Telegram for message: %s",
+                    user_text,
+                )
+            except asyncio.CancelledError:
+                logger.warning(
+                    "send_message cancelled (SDK cleanup leak) for message: %s",
                     user_text,
                 )
 

--- a/tests/test_claude_client.py
+++ b/tests/test_claude_client.py
@@ -455,3 +455,43 @@ async def test_send_closes_generator_on_normal_exhaustion():
     result = await client.send("test")
     assert result == "hello"
     assert closed, "async generator was not properly closed"
+
+
+# ---------------------------------------------------------------------------
+# send() — cancel scope cleanup mitigation (issue #33)
+# ---------------------------------------------------------------------------
+
+
+async def test_send_yields_control_after_stream(monkeypatch):
+    """Regression #33: send() must await asyncio.sleep(0) after the stream
+    closes so any leaked cancel-scope cleanup settles before the caller
+    does further I/O.
+    """
+    from claude_agent_sdk import AssistantMessage, TextBlock, ResultMessage
+
+    sleep_called_with = []
+    original_sleep = asyncio.sleep
+
+    async def spy_sleep(seconds, *args, **kwargs):
+        sleep_called_with.append(seconds)
+        return await original_sleep(seconds, *args, **kwargs)
+
+    monkeypatch.setattr("corphish.claude_client.asyncio.sleep", spy_sleep)
+
+    messages = [
+        AssistantMessage(content=[TextBlock(text="hi")], model="test"),
+        ResultMessage(
+            subtype="success",
+            duration_ms=100,
+            duration_api_ms=80,
+            is_error=False,
+            num_turns=1,
+            session_id="s1",
+        ),
+    ]
+    client = _make_client(query_fn=_make_query_fn(messages))
+    result = await client.send("test")
+    assert result == "hi"
+    assert 0 in sleep_called_with, (
+        "asyncio.sleep(0) was not called after stream closed"
+    )

--- a/tests/test_daemon.py
+++ b/tests/test_daemon.py
@@ -301,3 +301,58 @@ async def test_daemon_persists_offset_even_when_claude_fails():
     await run_daemon(**{k: v for k, v in deps.items() if k != "_bot"})
 
     deps["save_offset_fn"].assert_called_once_with(301)
+
+
+# --- CancelledError resilience tests (issue #33) ---
+
+
+async def test_daemon_continues_after_send_message_cancelled_error():
+    """Regression #33: CancelledError on send_message must not crash the daemon."""
+    updates = [
+        _make_update(1, 42, "first"),
+        _make_update(2, 42, "second"),
+    ]
+    deps = _make_deps(chat_id=42, updates=updates)
+    deps["claude"].send = AsyncMock(side_effect=["r1", "r2"])
+    deps["send_message_fn"] = AsyncMock(
+        side_effect=[asyncio.CancelledError("cancel scope leak"), None]
+    )
+
+    await run_daemon(**{k: v for k, v in deps.items() if k != "_bot"})
+
+    # Both messages should be processed; the daemon must not exit
+    assert deps["claude"].send.await_count == 2
+    assert deps["send_message_fn"].await_count == 2
+
+
+async def test_daemon_continues_after_claude_cancelled_error():
+    """Regression #33: CancelledError from Claude SDK must not crash the daemon."""
+    updates = [
+        _make_update(1, 42, "boom"),
+        _make_update(2, 42, "ok"),
+    ]
+    deps = _make_deps(chat_id=42, updates=updates)
+    deps["claude"].send = AsyncMock(
+        side_effect=[asyncio.CancelledError("cancel scope leak"), "reply-ok"]
+    )
+
+    await run_daemon(**{k: v for k, v in deps.items() if k != "_bot"})
+
+    assert deps["claude"].send.await_count == 2
+    # First message's send_message should be skipped; only second succeeds
+    deps["send_message_fn"].assert_awaited_once_with(deps["_bot"], 42, "reply-ok")
+
+
+async def test_daemon_logs_cancelled_error_on_send(caplog):
+    """CancelledError on send_message should be logged as a warning, not crash."""
+    updates = [_make_update(1, 42, "hi")]
+    deps = _make_deps(chat_id=42, updates=[_make_update(1, 42, "hi")])
+    deps["claude"].send = AsyncMock(return_value="reply")
+    deps["send_message_fn"] = AsyncMock(
+        side_effect=asyncio.CancelledError("cancel scope leak")
+    )
+
+    with caplog.at_level(logging.WARNING):
+        await run_daemon(**{k: v for k, v in deps.items() if k != "_bot"})
+
+    assert any("send_message cancelled" in r.message for r in caplog.records)

